rubygems-org-compact-index-api.md

```diff
@@ -0,0 +1,189 @@
+    repr-digest: sha-256="pf7Ts7NTRac//JxO9ke3IYbcWx8kcVn9N0mbm5EJpP0="
+    accept-ranges: bytes
+
+All compact index endpoints support ETags with the `If-None-Match` header.
+
+    If-None-Match: "40148273a7c7cd16b49d00a9935fd445"
+
+The compact index is designed to be fetched using the HTTP `Range` header.
+When a previously fetched copy is present, a ranged request to take advantage of the appended line pattern.
+
+    Range: bytes=#{range_start}-
+
+Responses from the `/versions` and `/info/[GEM]` endpoints will include the `Repr-Digest` header.
+The digests will have at least the SHA256 checksum of the entire file whether the response includes the full file or a partial response.
+*Please note that a `Digest` header is present, but it is deprecated and may be removed without notice.*
+
+When the Range header is satisfied, append the contents at exactly the starting byte, then computer the SHA256 checksum of the resulting file.
+If the result matches the `Repr-Digest` header's SHA256 checksum, the file is considered complete and up-to-date.
+
+Each line in the `/versions` file includes the latest MD5 calculation of the matching `/info` file for that gem at the time the line was written.
+The latest MD5, closest to the end of the file, will match the MD5 of the up-to-date `/info` file.
+
+Exact implementation details are available in the [Bundler CompactIndexClient::Updater](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/compact_index_client/updater.rb).
+
+### GET - `/versions`
+
+Returns a custom text based format of all versions of all gems.
+
+This API endpoint is intended to be a compact index of all possible gem versions.
+When new gems are added or when a gem is yanked, it is added to the end of the file using a specific format detailed below.
+
+The file is append only during the month which improves caching performance.
+It is recalculated at the start of each month, removing yanked gems and adding any new versions into the line for that gem.
+
+*Warning: this is a big file. Example is truncated.*
+
+    $ curl https://rubygems.org/versions
+    
+    created_at: 2024-04-01T00:00:05Z
+    ---
+    - 1 05d0116933ba44b0b5d0ee19bfd35ccc
+    -A 0.0.0 8b1527991f0022e46140907a7fc4cfd4
+    .cat 0.0.1 631fd60a806eaf5026c86fff3155c289
+    .omghi 1,2 7a67c0434100c2ab635b9f4865ee86bd
+    0mq 0.1.0,0.1.1,0.1.2,0.2.0,0.2.1,0.3.0,0.4.0,0.4.1,0.5.0,0.5.1,0.5.2,0.5.3 6146193f8f7e944156b0b42ec37bad3e
+    [...]
+    active_model_serializers -0.9.10 7ad37af4aec8cc089e409e1fdec86f3d
+    active_model_serializers 0.9.11 a6d40e97b289ee6c806e5e9f7031623b
+    openapi_first 1.4.1 40fbfdebcbfee3863df697e1d641f637
+
+#### `versions` File Format
+
+The format of the `versions` file uses one line per gem at computation time, with additional lines appended to the end that may include new or yanked versions of a gem already present earlier in the file.
+
+The first line is a created_at timestamp in IOS8601 format.
```

Comment:
I think for all compact index files, everything up to the `---` is opaque metadata. Furthermore, the opaque metadata and `---` separator line are optional.

Reply:
If the dashes are optional and your are reading the file, you can't look for the `---` as the start of the parseable data. Do you just try to read lines until you can parse one? It seems hard to parse reliably if the `---` is optional.

Reply:
https://github.com/rubygems/rubygems/blob/3bcc642d7127cc717c11ebff68d2b459a50e6e4b/bundler/lib/bundler/compact_index_client/cache.rb#L97-L102